Add documentation for instrumentation logic (#27918)
gradle/gradle: platforms/core-runtime/INSTRUMENTATION.md

> # Instrumentation and bytecode upgrades infrastructure

Reviewer: :x: I think "instrumentation" is used confusingly for multiple purposes, and it would be important to use distinct names for the different aspects here. Here's a proposal:

- what CC does is called "bytecode instrumentation"
- what we do for provider API migration is "bytecode upgrades"
- the infrastructure serving these both is "bytecode interception"

WDYT?

Author: I am ok with first two, but 3rd is similar to instrumentation.

What about Bytecode transformation for the infrastructure. So "bytecode transformation" infrastructure.

Author: Hmm, or maybe we can go with interception. Let's try it 🤔

Author: I now use "interception" for general term when talking about bytecode replacements. And I use "transform" when I am talking about transforming jars or classpath or classes.

## Abstract

This README describes the current state instrumentation and bytecode infrastructure used for configuration cache instrumentation and byte code upgrades used for property migration.

## Background

With the introduction of configuration cache we got a requirement to track user inputs for the configuration phase. 
Inputs could be environment variables, files, system properties and so on. 
These inputs are often read via an API that Gradle doesn’t control, e.g. Java API via System.getenv().

> So to track these inputs we had to find a way to track these usages. 

Reviewer: ```suggestion
So to track these inputs we had to find a way to track when they are accessed by intercepting API calls reading them. 
```

And that is why we started instrumenting bytecode.

> With a requirement to migrate Gradle core APIs to lazy properties, we were looking at how we could make old plugins compatible with new Gradle core APIs. 

Reviewer: ```suggestion
Later, with a requirement to migrate Gradle core APIs to lazy properties, we were looking at how we could make old plugins compatible with new Gradle core APIs. 
```

We decided that upgrading old compiled bytecode would work well. And reusing configuration cache bytecode instrumentation seemed like a good tool to reuse.

So configuration cache instrumentation and API upgrades are using the same infrastructure now.

## Functional design

> Imagine we have in Gradle core a task:

Reviewer: ```suggestion
Let's look at how bytecode interception works on an example of upgrading a JavaBean property to a lazy one.

Imagine we have in Gradle core a task:
```

> ```

Reviewer: :x: Let's add the languages here so that we get syntax highlighting.

Author: Done

abstract class JavaCompile {
    
    private String sourceCompatibility = null;
    
    @Input
    public String getSourceCompatibility() {
        return sourceCompatibility;
    }
    
    public void setSourceCompatibility(String sourceCompatibility) {
        this.sourceCompatibility = sourceCompatibility;
    }
} 
```
and we want to change it to:

```
abstract class JavaCompile {
    @Input
    public Property<String> getSourceCompatibility();
} 
```

To make old plugins compatible we need to replace all the usages of old JavaCompile with the new one.

Let’s say it’s used in a plugin as:

```
fun applyPlugin(project: Project) {
     project.tasks.named("compileJava", JavaCompile.class, task -> {
	  task.setSourceCompatibility("1.8");
     });
}
```

We will then replace all such usages with bytecode instrumentation with:
```
fun applyPlugin(project: Project) {
     project.tasks.named("compileJava", JavaCompile.class, task -> {
	    JavaCompileInterceptorsDeclaration.intercept_setSourceCompatibility(task, "1.8");
     });
}
```

> Where JavaCompileInterceptorsDeclaration is a class defined by Gradle developer to replace original logic and the replacement is done via bytecode instrumentation of plugin jars and dependencies.

Reviewer: ```suggestion
Where `JavaCompileInterceptorsDeclaration` is a class defined by a Gradle developer to replace the original logic, and the replacement is done via bytecode instrumentation of plugin jars and dependencies.
```

Due to that Instrumentation and bytecode upgrades is split into two big parts:
1. Gradle distribution build time: Interceptor code generation and Gradle API type metadata collection
2. Gradle runtime: Actual bytecode instrumentation

> JavaCompileInterceptorsDeclaration is defined by Gradle developer and with help of annotation processor Gradle then generates bytecode replacements. 

Reviewer: ```suggestion
`JavaCompileInterceptorsDeclaration` is defined by Gradle developer and with help of annotation processor Gradle then generates bytecode replacements. 
```

The actual replacement happens when the user runs his Gradle build.

**_Note: Groovy interception uses the same mechanism but replacements are a bit more complicated due to the dynamic nature of Groovy. 

> And due to that build logic is also intercepted when code is run and not just on the bytecode level._**

Reviewer: 🤔 I think dynamic Groovy is important enough to warrant its own section, WDYT?

Also: we should clarify that this is only about dynamic Groovy.

Author: Added a few sentences about dynamic Groovy and also pointed out that this is about dynamic Groovy

Next sections will explain more in detail how we can define interceptors and how the code is then replaced.

### Gradle Distribution Build Time

At Gradle Distribution Build time we have two major parts:

> 1. Declaring interceptors and generating interceptors

Reviewer: ❓ Declaring the interceptors happens before build time, i.e. when the developer adds the annotations to the source code, no? Shouldn't this say instead something like "we generate interceptors _from the annotations declared in the source code?_

Collecting the Gradle API type hierarchy is fully automatic, while declaring interceptors have to be done by Gradle developers. 

```mermaid
flowchart TB
    api[[Gradle API]] --> transform{{"Gradle API type\nhierarchy collector\n(Artifact transform)"}}
    declaredInterceptors[[Interceptors declarations]] --> processor{{"Interceptor generator\n(Annotation processor)"}}
    processor --> generatedInterceptors[[Generated Interceptors]]
    transform --> metadata[[Gradle API type metadata]]
    metadata --> distribution(Gradle distribution)
    implementation[[Gradle implementation]] --> distribution
    generatedInterceptors --> distribution
```

#### Declaring interceptors and generating interceptors

##### Annotation processor

Annotation processor is implemented as Java annotation processor in [platforms/core-runtime/internal-instrumentation-processor](https://github.com/gradle/gradle/tree/master/platforms/core-runtime/internal-instrumentation-processor) and its API (annotations) are defined in the [platforms/core-runtime/internal-instrumentation-api](https://github.com/gradle/gradle/tree/master/platforms/core-runtime/internal-instrumentation-api). 

> This annotation processor transforms interceptor declarations defined by Gradle developers to interceptors that modify Jvm bytecode.

Reviewer: :x: We shouldn't link using an absolute reference like this. Keep in mind that the structure might change etc., so "linking across time" (which is what you do when you link to `master`) can easily become confusing. I wonder if we can link to subdirectories with a relative path somehow? If not, then I think just mentioning their names is better.

Author: Makes sense, used relative paths

> Since this is a Java annotation processor all declarations have to be defined as Java classes, so interceptor declarations in Kotlin or Groovy are not supported.

Reviewer: I think this is important enough to call a bit more attention to it.

```suggestion
**Note:** Since this is a Java annotation processor all declarations have to be defined as Java classes, so interceptor declarations in Kotlin or Groovy are not supported.
```

8. **Logic** with call interception logic or logic that translates the new type to the old type

This JavaCompileInterceptorsDeclaration will then generate interceptors in **org.gradle.internal.classpath.generated** package.

For Java:
```
public class InterceptorDeclaration_JvmBytecodeCallInterceptor 
       extends MethodVisitorScope implements JvmBytecodeCallInterceptor {

     // Skipped fields and constructor for brevity
    @Override
    public boolean visitMethodInsn(String className, int opcode, String owner, String name,
            String descriptor, boolean isInterface, Supplier<MethodNode> readMethodNode) {
        if (owner.equals("org/gradle/api/tasks/compile/JavaCompile")) {
            if (name.equals("setSourceCompatibility") 
                    && descriptor.equals("(Ljava/lang/String;)V") 
                    && opcode == Opcodes.INVOKESPECIAL) {
                 // Skip some bytecode for brevity
                _INVOKESTATIC("JavaCompileInterceptorsDeclaration", "intercept_setSourceCompatibility", "(Lorg/gradle/api/tasks/compile/JavaCompile;Ljava/lang/String;)V");
                return true;
            }
       }
    }
}
```

> For Groovy:

Reviewer: ```suggestion
For dynamic Groovy:
```

```
public static class SetSourceCompatibilityCallInterceptor extends CallInterceptor {
    public SetSourceCompatibilityCallInterceptor() {
        super(InterceptScope.methodsNamed("setSourceCompatibility"));
    }

    @Override
    public Object doIntercept(Invocation invocation, String consumer) throws Throwable {
        Object receiver = invocation.getReceiver();
        if (receiver instanceof JavaCompile) {
            File receiverTyped = (JavaCompile) receiver;
            if (invocation.getArgsCount() == 1) {
                return JavaCompileInterceptorsDeclaration.intercept_setSourceCompatibility(
                        receiverTyped, 
                        (String) invocation.getArgument(0)
                );
            }
        }
        return invocation.callOriginal();
    }
}
```

> These two classes are then used when we instrument bytecode for Java and when we intercept calls at runtime for Groovy.

Reviewer: ```suggestion
These two classes are then used when we instrument bytecode for Java and when we intercept calls at runtime for dynamic Groovy.
```
